### ISSUE TITLE
cola.drag() throws errors with d3v7

### ISSUE THREAD

**Issue by l0drex** [tgdwyer/WebCola#326]
_selection_.call(cola.drag) throws a bunch of errors when using the latest d3 version, d3v7. When using cola.js (not the min version), the output is as follows:

```
Uncaught TypeError: position is undefined
    drag layout.js:532
    drag d3v4adaptor.js:31
    call d3.v7.js:1431
    gesture d3.v7.js:2665
    mousemoved d3.v7.js:2588
    contextListener d3.v7.js:2215
    onAdd d3.v7.js:2254
    selection_each d3.v7.js:1889
    selection_on d3.v7.js:2277
    mousedowned d3.v7.js:2572
    contextListener d3.v7.js:2215
    onAdd d3.v7.js:2254
    selection_each d3.v7.js:1889
    selection_on d3.v7.js:2277
    drag d3.v7.js:2558
    selection_call d3.v7.js:1855
    drag d3v4adaptor.js:38
    selection_call d3.v7.js:1855
```

**Comment by ViggoV** [comment]
Did you find a solution for this?

**Comment by l0drex** [comment]
No, I'm not using the drag function anymore. You could try to write your own adaptor like [this one](https://github.com/tgdwyer/WebCola/blob/master/src/d3v4adaptor.ts)

**Comment by ViggoV** [comment]
Thanks. I'll look into it :)

**Comment by JustyrTsaiLar** [comment]
v7 same problem .

how to make a my own adaotor ? still no drag moving and no error console .
is drag from d3.on or cola.drag ? 
and event name using like start.adptor or start ?

thanks

**Comment by ViggoV** [comment]
> v7 same problem .
> 
> how to make my own adaotor ? still no drag moving and no error console . is drag from d3.drag or cola.drag ? and event name using like start.adptor or start ?
> 
> thanks

I ended up copying the d3v4 adapter linked above and simplifying it a bit. Turned out I could avoid using the internal imports (at least for my use case) with a few tweaks, and by converting it to JS so the TypeScript time were redundant.

[Here is a Gist](https://gist.github.com/ViggoV/09f8b5639968a8252f3e6fafa6ad9298)

**Comment by cbbeau** [comment]
This was really helpful! To make this work in my case, I also needed to modify the calls to dragStart, dragOrigin, and dragEnd, not just drag. Example for dragStart:

`.on('start.d3adaptor', Layout.dragStart)`

to 

`.on("start.d3adaptor", (e, d)=>{ Layout.dragStart(d); })`

because the dragStart function expects only d (the node or group) but was receiving the event instead.
